### PR TITLE
Refactor index string fallbacks for Contributor Index

### PR DESCRIPTION
**Pull request by bryophyta**
## What does this change?

Replaces a chain of `orElse` statements with string concatenation. The previous structure didn't test whether `tags.lastName` was an empty string. Passing an empty string to `tagPages.alphaIndexKey` doesn't crash the job, but it [was leading to an error being logged](https://github.com/guardian/frontend/issues/25785#issuecomment-1366853945).

@jamesgorrie's PR #25832 will downgrade the error message to an info message. 

This PR should remove a significant proportion of the events which were triggering the logging statement in the first place. Using string concatenation here means that we don't need to explicitly test whether `lastName` is empty; if it _is_ empty then the first character in `firstName` will be used for the index (and likewise if `firstName` is also empty). This is a less 'lazy' approach as it will often involve evaluating values that don't get used, but the cost of eagerness here seems low because it just involves unwrapping some Options and concatenating three strings.

This PR should also affect the contributor indexes that are built by this job, because all the contributors with no `lastName` should now be indexed by their `firstName` or `webTitle` instead of the fallback index key of `1-9`. It looks like this was the intention of the original code, and in theory should be a more 'useful' index. However, it will still provide the 'wrong' bucketing and alphabetical ordering for those contributors who have no `lastName` set through human error (as opposed to the case where their name doesn't fit within the convention of a first and last name; in the latter cases ideally the affordances provided here would be used by editorial to produce the conventionally correct ordering based on the specifics of the case).

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

## Tests

I deployed this to CODE last night and the errors seem to have dropped off since that deployment:
image

Also, @ioannakok and I have verified that when the job runs with the code in this PR it does rebuild the indexes in S3 with contributors allocated as we'd expect (i.e. most of them have moved from [the `1-9` bucket](https://www.theguardian.com/index/contributors/1-9) into an alphabet bucket):

|Before|After|
|--|--|
|the 1-9 page only has many entries for contributors: Frank, David Aaronovitch, Abahachi, etc.|the 1-9 page only has one entry now: Studio 20 NYU|

### CODE REVIEW

File: admin/app/jobs/RebuildIndexJob.scala [hunk 1]
```diff
@@ -96,7 +96,16 @@ class RebuildIndexJob(contentApiClient: ContentApiClient)(implicit executionCont
...
       val contributorsByNameOrder: Map[String, Set[Tag]] = contributors.toSet
-        .groupBy(tag => tagPages.alphaIndexKey(tag.lastName orElse tag.firstName getOrElse tag.webTitle))
+        .groupBy(tag => {
+
+          /**
+            * The alphaIndexKey function looks at the first alphanumeric character in the string passed to it.
+            * Concatenating these three strings therefore allows us to use firstName and webTitle as fallbacks
+            * if lastName is None or "".
+            * */
+          val indexString = tag.lastName.getOrElse("").trim + tag.firstName.getOrElse("").trim + tag.webTitle
```
Comment (OllysCoding) [inline admin/app/jobs/RebuildIndexJob.scala:106]: Is there a risk of these being duplicated between contributors? If so do we know the effect that this might have?

Comment (bryophyta) [reply admin/app/jobs/RebuildIndexJob.scala:106]: good question! my understanding is that these values are just used to bucket contributors by the first letter in their names, if available. (More context on this functionality can be found in a previous PR from @ioannakok: #25176) So there shouldn't be any need for the `indexString` value to be unique, if I've understood the functionality correctly.

**Comment by ioannakok** [comment]
Perhaps this is a separate issue and I'm misunderstanding something here but even in the current situation I am not sure why contributors with both first name and last name end up in the `1-9` index: https://www.theguardian.com/index/contributors/1-9

**Comment by bryophyta** [comment]
> Perhaps this is a separate issue and I'm misunderstanding something here but even in the current situation I am not sure why contributors with both first name and last name end up in the `1-9` index: [theguardian.com/index/contributors/1-9](https://www.theguardian.com/index/contributors/1-9)

It's a good question! I'm not 100% sure but I _believe_ it's generally going to be because their information has been put in `wrong`, so that for instance the `webTitle` has the full name, but the `lastName` field has been left blank. e.g. the 1-9 page includes `David Aaronovitch`, but if you look at the CAPI contributor response we see that there's no `lastName` set for this contributor:

image

**Comment by ioannakok** [comment]
> > Perhaps this is a separate issue and I'm misunderstanding something here but even in the current situation I am not sure why contributors with both first name and last name end up in the `1-9` index: [theguardian.com/index/contributors/1-9](https://www.theguardian.com/index/contributors/1-9)
> 
> It's a good question! I'm not 100% sure but I _believe_ it's generally going to be because their information has been put in "wrong", so that for instance the `webTitle` has the full name, but the `lastName` field has been left blank. e.g. the 1-9 page includes `David Aaronovitch`, but if you look at the CAPI contributor response we see that there's no `lastName` set for this contributor:
> 
> image

Ah, I see - thank you!

File: admin/app/jobs/RebuildIndexJob.scala (hunk 1 above)
Comment (ioannakok) [inline admin/app/jobs/RebuildIndexJob.scala:106]:
```suggestion
          val indexString = tag.lastName.getOrElse("").trim.concat(tag.firstName.getOrElse("").trim).concat(tag.webTitle)
```

Comment (ioannakok) [inline admin/app/jobs/RebuildIndexJob.scala:106]:
```suggestion
          val indexString = tag.lastName.getOrElse("").trim.concat(tag.firstName.getOrElse("").trim).concat(tag.webTitle)
```

Comment (bryophyta) [reply admin/app/jobs/RebuildIndexJob.scala]: thanks! I committed this locally change so that I could run `scalafmt` 👍

**Review by ioannakok — APPROVED** [review]
Looks good to me!

**Review by jamesgorrie — COMMENTED** [review]
This looks good - but just had a thought about making these indexes simpler in general. Could we just use the `webTitle`?

Looking at the `Tag` model - [webTitle is required](https://github.com/guardian/content-api-models/blob/master/models/src/main/thrift/content/v1.thrift#L1224) and almost always seems to be some combination of names. 

Looking at where we [render the text](https://github.com/guardian/frontend/blob/main/applications/app/views/fragments/tagIndexBody.scala.html#L19), it is also [the webTitle with some sort of deduping](https://github.com/guardian/frontend/blob/main/common/app/model/TagIndexPage.scala#L93-L96).

**Comment by jamesgorrie** [comment]
>  then we order them alphabetically by last name. 

I would love to speak to the person who thought an index sorted on multiple, unsurfaced criteria would be easier to search. Saying that, it's a good point of not knowing the intent.

Let's go digging for that.

I am going to take a guess it was:
* You normally sort by last name, let's do that
* Shizzle, we don't always have that, let's fallback to `firstname`
* Oh my, sometimes that's missing, fallback to `webTitle`
